### PR TITLE
ARROW-8295: [C++][Dataset] Push down projection to IpcReadOptions

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

**Comment by nealrichardson** [comment]
@bkietz this needs rebase

**Comment by nealrichardson** [comment]
CI is green, just need to resolve the merge conflict and then we can merge--any objections?

**Review by wesm — COMMENTED** [review]
This looks OK to me. Can you open a JIRA about avoiding redundant schema munging in the stream and file readers? We are doing this for every record batch and preferably we should only do it once.

**Comment by wesm** [comment]
+1. I'll go ahead and merge this

File: c_glib/test/test-feather-file-reader.rb [hunk 1]
```diff
@@ -65,7 +65,7 @@ def setup_file(table)
     setup_file(table) do |reader|
       assert_equal(build_table("message" => build_string_array(["Login"]),
                                "host" => build_string_array(["www"])),
-                   reader.read_names(["host", "message"]))
+                   reader.read_names(["message", "host"]))
```
[the remark this answers is not in the document]
Comment (bkietz) [reply c_glib/test/test-feather-file-reader.rb:68]: Yes

**Comment by bkietz** [comment]
https://issues.apache.org/jira/browse/ARROW-8434

**Pull request closed without merging**